### ISSUE TITLE
[HOLD for payment 2023-10-24] [$500] Workspace - Admin Checkbox  in Members Section Ticked When Changing Contact Method

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#28539]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to Settings > Profile > Contact Method > Add New Contact Method
2. Go to Settings > Create Workspace > Members > Invite Members
3. Go to Settings > Profile > Contact Method > Change Contact Method
4. Go to Settings > Workspace > Members > Tick on the Select All checkboxes and observe that the Admin checkbox is also selected

## Expected Result:
Changing the contact method should not cause the Admin checkbox to be ticked in the Members section.

## Actual Result:
Changing the contact method causes the Admin checkbox to be ticked in the Members section.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.73.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/1f95a553-1a76-4b0d-9e56-3ffc2b0cc84b

https://github.com/Expensify/App/assets/93399543/fd16dd74-306d-40aa-912c-7d742fd0433d

https://github.com/Expensify/App/assets/93399543/e73b3e91-1a27-4ffe-97c7-583fadd91ead

https://github.com/Expensify/App/assets/93399543/899aea1d-4981-4c94-819e-4b3f4de6b61e

https://github.com/Expensify/App/assets/93399543/3a63913c-a900-48d5-b3f1-5c4f36296a99

[screen-capture - 2023-09-22T140546.004.webm](https://github.com/Expensify/App/assets/93399543/40887850-3101-42d9-a9e1-1ae446b43c47)

https://github.com/Expensify/App/assets/93399543/6a0163d5-555d-4367-bfec-7a461ddbd2c2

**Expensify/Expensify Issue URL:**
**Issue reported by:** @tewodrosGirmaA
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1695415622144909

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by esh-g** [comment]
# Proposal

## Please re-state the issue
Able to select the policy owner after changing the contact method.

## What is the root cause of this issue?
While checking whether an option should be disabled, we are setting the `isDisabled` property to true here:
https://github.com/Expensify/App/blob/4ea4a55ad2c9ecdb66d45cc1872612ea22e4605f/src/pages/workspace/WorkspaceMembersPage.js#L317

But the `accountID` from `props.policyMembers` is actually a `string` while the `props.session.accountID` is actually a `Number`.

## What changes do you think should be made to fix this problem?
We should convert the `accountID` from `props.policyMembers` to `Number` before comparing like we do in many other places for `accountID`.
```js
isDisabled: Number(accountID) === props.session.accountID || details.login === props.policy.owner || policyMember.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE,
```

## What other alternative options did you explore? (Optional)

**Comment by alexpensify** [comment]
@parasharrajat - when you get a chance can you review if one of these proposals will resolve this issue? Thanks!

**Comment by alexpensify** [comment]
@parasharrajat any update here?

**Comment by alexpensify** [comment]
Marc is OOO but will review soon. If Marc is unable to help, I'll find help to keep it moving forward.

**Comment by marcaaron** [comment]
>the accountID from props.policyMembers is actually a string while the props.session.accountID is actually a Number.

Not sure why the `accountID` would ever be a string though. This sounds like more of a backend error to me.

cc @puneetlath I think this change was first introduced [here](https://github.com/Expensify/App/pull/21237) in case you have some idea.

**Comment by marcaaron** [comment]
Rejected the proposal because:

>We should convert the accountID from props.policyMembers to Number before comparing like we do in many other places for accountID.

It looks like we already do this [here](https://github.com/Expensify/App/blob/11acdc2a4918859a3e1b8dfe23f10ad9a6f24218/src/pages/workspace/WorkspaceMembersPage.js#L317-L318).

**Comment by marcaaron** [comment]
Passing the 🏀 back to you @parasharrajat 😄

**Comment by marcaaron** [comment]
Oh duh I see now that the isDisabled logic is set in the same object 🤦

**Comment by marcaaron** [comment]
I think ideally we can fix this at a lower level instead of cast to a `Number` everywhere. Let's find out where we are setting the `ONYXKEYS.COLLECTION.POLICY_MEMBERS` to a string.

**Comment by esh-g** [comment]
@marcaaron Actually the structure we are using right now for `policyMembers` is like this:
```js
{
  123: {role: 'admin'},
  345: {role: 'member'},
  ...
}
```
This means that all `accountIDs` are actually a key on the object and in javascript, they are implicitly casted to a string.

<img width="169" alt="Screenshot 2023-10-11 at 3 59 05 PM" src="https://github.com/Expensify/App/assets/77237602/c2a54949-2bc0-4464-81d5-60b7f07ad248">

cc @parasharrajat

**Comment by puneetlath** [comment]
>cc @puneetlath I think this change was first introduced https://github.com/Expensify/App/pull/21237 in case you have some idea.

I have no recollection! Perhaps just an oversight? I know reportIDs need to be strings, so maybe we just got mixed up.

**Comment by marcaaron** [comment]
@esh-g Ah!!! 😄 

Ok, new plan (that is basically the original plan) - and only cast once after here...
 https://github.com/Expensify/App/blob/4ea4a55ad2c9ecdb66d45cc1872612ea22e4605f/src/pages/workspace/WorkspaceMembersPage.js#L266

We can remove the other inconsistent casts then? Assuming the other accountID are not also strings 😮‍💨

**Comment by esh-g** [comment]
@marcaaron so you are saying that we should only once cast the accountID to a number and store it in a var like accountIDNum, right?

**Comment by esh-g** [comment]
@parasharrajat PR is ready: https://github.com/Expensify/App/pull/29422 💫

**Comment by marcaaron** [comment]
Yep exactly

**Comment by alexpensify** [comment]
Awesome, looks like the PR is moving forward!

**Comment by alexpensify** [comment]
@parasharrajat - to prepare for the payment date, can you please complete the checklist? Thanks!

**Comment by alexpensify** [comment]
@parasharrajat - please complete the checklist. Thanks!

**Comment by alexpensify** [comment]
@marcaaron - can we confirm if you agree with the regression test above?

**Comment by marcaaron** [comment]
LGTM

**Comment by alexpensify** [comment]
Here is the payment summary: 

-  External issue reporter @tewodrosGirmaA $50 
-  Contributor that fixed the issue @esh-g $750
-  Contributor+ that helped on the issue and/or PR @parasharrajat $750

Upwork Job: https://www.upwork.com/jobs/~01ccfbef1a9c9e1edd

*If applicable, the bonuses will be applied on the final payment 

Extra Notes regarding payment: There is an urgency bonus and hit the 3-day business day mark.

**Comment by alexpensify** [comment]
@tewodrosGirmaA and @esh-g - can you please apply to the link above and I can complete the next steps in Upwork. Not sure what happened but the automation didn't kick in here. Thanks!

**Comment by esh-g** [comment]
Applied @alexpensify 🚀

**Comment by tewodrosGirmaA** [comment]
Applies @alexpensify

**Comment by alexpensify** [comment]
Thanks, I need your approvals now and then can complete the process.

**Comment by esh-g** [comment]
Accepted the offer @alexpensify

**Comment by alexpensify** [comment]
Everyone has been paid via Upwork.

@parasharrajat please submit your request too. Thanks!

**Comment by alexpensify** [comment]
Alright, the regression test has been created so I'm going to close out the issue. Great work here!

**Comment by JmillsExpensify** [comment]
$750 payment approved for @parasharrajat based on [this comment](https://github.com/Expensify/App/issues/28539#issuecomment-1778243048).
